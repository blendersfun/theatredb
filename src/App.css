.app {
  padding: 20px;
}
.navMenu {
  display: flex;
  padding: 10px 20px;
  background: #eee;
}
.navLinks {
  flex: 1;
}
.navLinks > a + a {
  margin-left: 10px;
}
.whoami {
  padding-right: 20px;
}
.page {
  margin: 40px;
}