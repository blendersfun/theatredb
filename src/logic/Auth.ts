
import { client } from './stitch'
import {
  AnonymousCredential,
  UserPasswordCredential
} from 'mongodb-stitch-core-sdk'

export type User = {
  email: string
}

export class Auth {
  static async init(): Promise<void> {
    // Resolve stitch client auth before loading any pages:
    if (!client.auth.isLoggedIn) {
      await client.auth.loginWithCredential(new AnonymousCredential())
    }
    window.dispatchEvent(new CustomEvent('loginStateChanged'))
  }
  static async login(email: string, password: string): Promise<void> {
    await client.auth.logout()
    const credential = new UserPasswordCredential(email, password)
    await client.auth.loginWithCredential(credential)
    window.dispatchEvent(new CustomEvent('loginStateChanged'))
    window.dispatchEvent(new CustomEvent('navigate', {
      detail: { page: 'home' }
    }))
  }
  static async logout(): Promise<void> {
    await client.auth.logout()
    await client.auth.loginWithCredential(new AnonymousCredential())
    window.dispatchEvent(new CustomEvent('loginStateChanged'))
  }
  static user(): User|null {
    const user = client.auth.user
    if (!user) throw new Error('Logged in, but user was not defined.')

    // The typescript definitions appear to be out of data with what the service
    // is returning. Todo: file a bug about this.
    const profile: any = user.profile
    if (user.loggedInProviderType === 'anon-user') {
      return null
    }
    return {
      email: profile.data.email || null
    }
  }
}
