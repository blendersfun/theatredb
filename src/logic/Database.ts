import { db } from './stitch'

export class Database {
  static async fetch(collection: string): Promise<any> {
    return await db.collection(collection)
      .find({}, { limit: 1000 })
      .asArray()
  }
  static async add(collection: string, document: any): Promise<void> {
    await db.collection(collection).insertOne(document)
  }
}
