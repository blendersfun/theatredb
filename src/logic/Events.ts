const registry: any = {}
const callbacksByBind: any = {}

export class Events {
  static attach(eventName: string, callback: EventListener, bind: Object) {
    // Remember what callbacks go with this bind:
    const bindAsIndex = bind as any
    if(!callbacksByBind[bindAsIndex]) callbacksByBind[bindAsIndex] = []
    callbacksByBind[bindAsIndex].push(callback)

    // Remember what eventName and boundCallback go with this callback:
    const callbackAsIndex = callback as any
    const boundCallback: EventListener = callback.bind(bind)
    registry[callbackAsIndex] = { eventName, boundCallback }

    window.addEventListener(eventName, boundCallback)
  }
  static detatch(what: Object|EventListener) {
    if (typeof what === 'function') {
      const callbackAsIndex = what as any
      const entry: any = registry[callbackAsIndex]
      if (entry && entry.eventName && entry.boundCallback) {
        window.removeEventListener(entry.eventName, entry.boundCallback)
      }
    } else {
      const bindAsIndex = what as any
      const entries = callbacksByBind[bindAsIndex]
      if (entries instanceof Array) {
        for (let i = 0; i < entries.length; i++) {
          const callbackAsIndex = entries[i] as any
          const entry: any = registry[callbackAsIndex]
          if (entry && entry.eventName && entry.boundCallback) {
            window.removeEventListener(entry.eventName, entry.boundCallback)
          }
        }
      }
    }
  }
  static dispatch(eventName: string, detail?: any) {
    window.dispatchEvent(new CustomEvent(eventName, { detail }))
  }
}
