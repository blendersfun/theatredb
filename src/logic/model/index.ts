export { Organization } from './Organization'
export { Person } from './Person'
export { Production } from './Production'
export { Script } from './Script'
