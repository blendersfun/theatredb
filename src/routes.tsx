import React, { ReactElement } from 'react'

import { Home } from './pages/home/Home'
import { Login } from './pages/login/Login'
import { DataEntry } from './pages/data-entry/DataEntry'

export type RouteParams = {
  page: string
}

export type RouteConfig = {
  [page: string]: (params: RouteParams) => ReactElement<any>|null
}

export class Routes {
  static routes: RouteConfig = {
    'home': () => <Home/>,
    'login': () => <Login/>,
    'data-entry': () => <DataEntry/>
  }
  static evaulateRoute(routeParams: RouteParams): ReactElement<any> {
    for (const pageName of Object.keys(this.routes)) {
      if (pageName === routeParams.page) {
        const page = this.routes[pageName](routeParams)
        if (page) return page
      }
    }
    return <div>Error: No page found.</div>
  }
  static exists(routeName: string): boolean {
    return this.routes.hasOwnProperty(routeName)
  }
}
