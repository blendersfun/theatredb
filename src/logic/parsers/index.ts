import { Production, Script } from '../model'
import { PersonInRole } from '../model/Summaries'
import { ScriptDetail } from '../model/Script'

/**
 * Parses a display formatted date into a Date object 
 * using the current locale's timezone offset.
 * 
 * Example Raw Record: 3/9/2019
 */
function parseDate(raw: string): Date|undefined {
  const trimmed = raw.trim()
  if (!trimmed) return
  let [month, day, year] = trimmed.split('/')
  month = month.padStart(2, '0')
  day = day.padStart(2, '0')
  const offset = Math.floor(new Date().getTimezoneOffset() / 60).toString().padStart(2, '0')
  const date = new Date(`${year}-${month}-${day}T19:00:00-${offset}00`)
  const dateFormatting = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' }
  console.log('parsedDate', date.toLocaleString('en-US', dateFormatting))
  return date
}

/**
 * Parses a list of authors into an array of Person objects.
 * 
 * Example Raw Records:
 *  - David Henry Hwang
 *  - Lauren Gunderson and Margot Melcon
 *  - Charles Dickens, adapted to the stage by Romulus Linney
 *  - based on the novel by Man Ray, adapted by Abraham Lincoln
 *  - adapted by Jeffrey Hatcher from the original by Nikolai Gogol
 *  - Molière, translated into English verse by Richard Wilbur
 */
function parseAuthors(raw: string): PersonInRole[] {
  if (!raw.trim()) return []
  const authors: PersonInRole[] = []
  let adaptor, novelist
  if (raw.includes(', translated into English verse by')) {
    [, novelist, adaptor] = /([\wè ]+), translated into English verse by ([\w ]+)/.exec(raw) || [,,,]
  } else if (raw.includes('from the original by')) {
    [, adaptor, novelist] = /adapted by ([\w ]+) from the original by ([\w ]+)/.exec(raw) || [,,,]
  } else if (raw.search(/, adapted (to|for) the stage by /) !== -1) {
    [, novelist,, adaptor] = /([\w ]+), adapted (:?to|for) the stage by ([\w ]+)/.exec(raw) || [,,,,]
  } else if (raw.includes('based on the novel')) {
    [, novelist, adaptor] = /based on the novel by ([\w ]+), adapted by ([\w ]+)/.exec(raw) || [,,,]
  } else {
    raw.trim().split(' and ').forEach(name => {
      authors.push({ person: { name }, role: 'playwright' })
    })
  }
  if (novelist) authors.push({ person: { name: novelist }, role: 'novelist' })
  if (adaptor) authors.push({ person: { name: adaptor }, role: 'playwright' })
  return authors
}

/**
 * Parses a list entry from Dramatist's Play Service's current productions list.
 * - Example URL: https://www.dramatists.com/dps/productions.aspx
 * - Example Raw Record:
 * 
 *   The Call (Barfield)
 *   Tanya Barfield
 *   Seattle Public Theatre  | Seattle,  WA  Professional production,  Opening: 5/10/2019 thru  6/9/2019)
 */
export function entryFromDPSList(raw: string): Production|null {
  if (!raw.trim()) return null
  
  const lines = raw.trim().split(/[\r\n]+/)
  const scriptName = lines[0].trim()
  const authors = parseAuthors(lines[1] || '')

  let [orgName, rest] = (lines[2] || '').split('|')
  rest = (rest || '').trim()
  const splits = rest.split(/\s+/)
  const city = splits[0].replace(',', '')
  const state = splits[1] || ''
  const isProfessional = splits[2] === 'Professional'
  const openingIndex = splits.indexOf('Opening:')
  const openingRaw = openingIndex === -1 ? '' : splits[openingIndex + 1]
  const closingIndex = splits.indexOf('thru')
  const closingRaw = closingIndex === -1 ? '' : splits[closingIndex + 1].replace(')', '')

  const production = {
    name: scriptName,
    organization: { name: orgName.trim() },
    script: {
      name: scriptName,
      authors
    },
    isProfessional,
    location: {
      city: city.trim(),
      state: state.trim()
    },
    schedule: {
      openingDate: parseDate(openingRaw),
      closingDate: parseDate(closingRaw)
    }
  }

  return new Production(production)
}

/**
 * Parses a script detail page from Dramatist's Play Service.
 * - Example Url: http://www.dramatists.com/cgi-bin/db/single.asp?key=3971
 * - Example Record:
 * 
 *   A Doll's House, Part 2
 *   Lucas Hnath
 *   ISBN: 978-0-8222-3744-0
 *   Full Length, Drama
 *   1 man, 3 women
 *   Total Cast: 4, Flexible Set
 *   $100 per performance.
 *   In the final scene of Ibsen’s 1879 groundbreaking masterwork, Nora Helmer makes the shocking decision to leave her husband and children, and begin a life on her own. This climactic event—when Nora slams the door on everything in her life—instantly propelled world drama into the modern age. In A DOLL’S HOUSE, PART 2, many years have passed since Nora’s exit. Now, there’s a knock on that same door. Nora has returned. But why? And what will it mean for those she left behind?
 *   “[A] smart, funny and utterly engrossing play…Hnath approaches what might seem like a hubristic project with the humility and avidity of an engaged Everyreader. A DOLL'S HOUSE, PART 2 gives vibrant theatrical life to the conversations that many of us had after first reading or seeing its prototype…” —NY Times. “…lucid and absorbing…Modern in its language, mordant in its humor and suspenseful in its plotting…the play judiciously balances conflicting ideas about freedom, love and responsibility.” —Time Out NY. “Hnath’s inspired writing, which endows each character with an arsenal of fastballs, curveballs and spitballs, keep[s] us disarmingly off-balance. He’s an uncommonly gifted parodist. For all its seriousness, A DOLL'S HOUSE, PART 2 is suffused with a contagious bemusement.” —Deadline.com. “[A DOLL'S HOUSE, PART 2] delivers explosive laughs while also posing thoughtful questions about marriage, gender inequality and human rights…as much an ingenious elaboration and deconstruction of A Doll’s House as a sequel, and it stands perfectly well on its own…With unfussy eloquence, [the play] asks how much, in a century-plus, life has changed for Nora and women like her in a world that often still has firm ideas about where they belong.” —Hollywood Reporter.
 *   {awards text would go here}
 */
export function entryFromDPSDetail(raw: string): Script|null {
  if (!raw.trim()) return null

  const lines = raw.trim().split(/\r|\n|\r\n/)
  const scriptName = lines[0].trim()
  const authors = parseAuthors(lines[1] || '')
  const isbn = (lines[2] || '').trim().replace('ISBN: ', '')

  const [, licenseFeePerPerformanceStr] = (lines[6] || '').match(/\$([\d\.]+)/) || [,,]
  const [, menCount] = (lines[4] || '').match(/(\d+) m[ae]n/) || [,,]
  const [, womenCount] = (lines[4] || '').match(/(\d+) wom[ae]n/) || [,,]
  const [, totalCastCount] = (lines[5] || '').match(/^Total Cast: (\d+)/) || [,,]

  let reviewsRaw = lines[8] ? lines[8].trim() : ''
  const reviews = []
  let safety = 0
  while(reviewsRaw.length) {
    safety += 1
    if (safety > 100) throw new Error('too much')
    const reviewStart = reviewsRaw.search(/[“"]/)
    if (reviewStart === -1) break
    reviewsRaw = reviewsRaw.slice(reviewStart + 1)
    const reviewEnd = reviewsRaw.search(/[”"]/)
    if (reviewEnd === -1) break
    const review = reviewsRaw.slice(0, reviewEnd)
    const [, source] = reviewsRaw.match(/\s+—([^\.“"]+)/) || [,,]
    if (review && source) {
      reviews.push({ review, source })
    }
    reviewsRaw = reviewsRaw.slice(reviewEnd + 1)
  }

  const roles: ScriptDetail["roles"] = {
    total: parseInt(totalCastCount || '-1')
  }
  const script: ScriptDetail = {
    name: scriptName,
    isbn,
    authors,
    roles
  }

  if (licenseFeePerPerformanceStr) {
    script.licenseFeePerPerformance = parseFloat(licenseFeePerPerformanceStr)
  }

  if (lines[7]) script.synopsis = lines[7].trim()

  if (menCount) roles.men = parseInt(menCount || '-1')
  if (womenCount) roles.women = parseInt(womenCount || '-1')

  if (reviews.length) {
    script.reviews = reviews
  }

  if (lines[9]) {
    script.awards = lines[9].trim()
  }

  return new Script(script)
}
