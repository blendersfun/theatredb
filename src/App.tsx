import './App.css'

import React, {
  Component,
  ReactElement
} from 'react'

import { RouteParams, Routes } from './routes'
import { Auth } from './logic/Auth'
import { Events } from './logic/Events'

export interface AppState {
  page: ReactElement<any>,
  isLoggedIn: boolean,
  userEmail: string|null
}
export class App extends Component<{}, AppState> {
  constructor(props: any) {
    super(props)
    this.state = {
      page: <div/>,
      isLoggedIn: false,
      userEmail: null
    }
  }
  async componentDidMount(): Promise<void> {
    this.attach()
    await Auth.init()
    this.navigate(this.routeParamsFromHash(window.location.hash))
  }
  componentWillUnmount() {
    Events.detatch(this)
  }
  attach() {
    Events.attach('hashchange', this.onHashChanged, this)
    Events.attach('loginStateChanged', this.onLoginStateChanged, this)
    Events.attach('navigate', this.onNavigate, this)
  }
  onNavigate(event: Event) {
    const customEvent = event as CustomEvent<RouteParams>
    this.navigate(customEvent.detail)
  }
  onHashChanged() {
    const routeParams = this.routeParamsFromHash(window.location.hash)
    this.setState({
      page: Routes.evaulateRoute(routeParams)
    })
  }
  onLoginStateChanged() {
    const user = Auth.user()
    this.setState({
      isLoggedIn: user !== null,
      userEmail: user && user.email
    })
  }
  routeParamsFromHash(hashStr: string): RouteParams {
    const hash = new URLSearchParams(hashStr.slice(1))
    let page = hash.get('page')
    if (!page || !Routes.exists(page)) page = 'home'
    return {
      page
    }
  }
  navigate(routeParams: RouteParams) {
    window.history.pushState({}, '', `#page=${routeParams.page}`)
    this.setState({
      page: Routes.evaulateRoute(routeParams)
    })
  }
  async logout(): Promise<void> {
    await Auth.logout()
  }
  render() {
    return (
      <AppUI {...this.state}
        logout={this.logout.bind(this)}
        />
    )
  }
}

export interface AppProps extends AppState {
  logout: () => void
}
export class AppUI extends Component<AppProps> {
  render() {
    return (
      <div className="App">
        <div className="navMenu">
          <div className="navLinks">
            <a href="#page=home">Home</a>
            <a href="#page=data-entry">Data Entry</a>
          </div>
          <div className="loginStatus">
            {this.renderLoginStatus()}
          </div>
        </div>
        {this.props.page}
      </div>
    )
  }
  renderLoginStatus(): ReactElement<any> {
    if (!this.props.isLoggedIn) {
      return <a href="#page=login">Login</a>
    } else {
      return <div>
        <span className="whoami">Logged in as {this.props.userEmail}</span>
        <a onClick={this.props.logout} href="javascript:;">Logout</a>
      </div>
    }
  }
}
