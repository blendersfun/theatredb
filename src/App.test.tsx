jest.mock('./logic/Auth')
jest.mock('./logic/Events')
jest.mock('./routes')

import React from 'react'
import renderer from 'react-test-renderer';
import { App, AppUI, AppProps } from './App'
import sinon from 'sinon'

import { Routes } from './routes'
import { Auth } from './logic/Auth'
import { Events } from './logic/Events'


describe('logic', function () {
  const init = Auth.init as jest.Mock
  const user = Auth.user as jest.Mock
  const evaluateRoute = Routes.evaulateRoute as jest.Mock
  const exists = Routes.exists as jest.Mock
  const attach = Events.attach as jest.Mock
  const detatch = Events.detatch as jest.Mock

  beforeEach(() => {
    sinon.stub(App.prototype, 'render').returns(<div>Fake App</div>)
    window.location.hash = ''
  })
  afterEach(() => {
    sinon.restore()
    init.mockClear()
    user.mockClear()
    evaluateRoute.mockClear()
    attach.mockClear()
    detatch.mockClear()
    window.location.hash = ''
  })
  it('initializes the state when constructed', () => {
    const app = new App({})
    expect(app.state.page).toMatchSnapshot()
    expect(app.state.isLoggedIn).toBe(false)
    expect(app.state.userEmail).toBe(null)
  })
  it('attaches events on mount and detatches them on unmount', done => {
    const testRender = renderer.create(<App/>)
    const app: App = testRender.getInstance() as any
    if (app === null) throw new Error('Expected app instance to be non-null.')
    setImmediate(() => {
      testRender.unmount()
      expect(Events.attach).toBeCalledTimes(3)
      expect(Events.detatch).toBeCalledTimes(1)
      done()
    })
  })
  it('redirects to page=home on mount when hash is empty, and initializes Auth state', done => {
    renderer.create(<App/>)
    setImmediate(() => {
      // Todo: refactor window.history into ./logic so it can be mocked easily
      // and verified here.
      expect(evaluateRoute).toBeCalledWith({ page: 'home' })
      expect(Auth.init).toBeCalledTimes(1)
      done()
    })
  })
  it('redirects to page=home on mount when hash is an unknown page', done => {
    window.location.hash = '#page=f10b42b0-9bae-47d6-9071-9d46ae917b5c'
    renderer.create(<App/>)
    setImmediate(() => {
      expect(evaluateRoute).toBeCalledWith({ page: 'home' })
      done()
    })
  })
  it('navigates to page=login on mount when hash is login', done => {
    window.location.hash = '#page=login'
    renderer.create(<App/>)
    exists.mockReturnValueOnce(true)
    setImmediate(() => {
      expect(evaluateRoute).toBeCalledWith({ page: 'login' })
      done()
    })
  })
  it('navigates to page=login when the hash updates to login', done => {
    const app: App = renderer.create(<App/>).getInstance() as any
    exists.mockReturnValueOnce(true)
    setImmediate(() => {
      window.location.hash = '#page=login'
      app.onHashChanged()
      expect(evaluateRoute).toBeCalledWith({ page: 'login' })
      done()
    })
  })
  it('navigates to page=login when a navigate event indicates to', done => {
    const app: App = renderer.create(<App/>).getInstance() as any
    exists.mockReturnValueOnce(true)
    setImmediate(() => {
      app.onNavigate(new CustomEvent('navigate', { detail: { page: 'login' } }))
      expect(evaluateRoute).toBeCalledWith({ page: 'login' })
      done()
    })
  })
  it('updates the user when a loginStateChanged event indicates to', done => {
    user.mockReturnValueOnce({ email: 'albus@hogwarts.edu' })
    const testRenderer = renderer.create(<App/>)
    const app: App = testRenderer.getInstance() as any
    app.onLoginStateChanged()
    setImmediate(() => {
      expect(user).toBeCalledTimes(1)
      expect(app.state.isLoggedIn).toBe(true)
      expect(app.state.userEmail).toBe('albus@hogwarts.edu')
      done()
    })
  })
})

describe('ui', () => {
  it('renders a test page with a user not logged in', () => {
    const props: AppProps = {
      page: <div> Test Page </div>,
      isLoggedIn: false,
      userEmail: null,
      logout: jest.fn()
    }
    expect(renderer.create(<AppUI {...props}/>).toJSON()).toMatchSnapshot()
  })
  it('renders a test page with a user logged in', () => {
    const props: AppProps = {
      page: <div> Test Page </div>,
      isLoggedIn: true,
      userEmail: 'snape@hogwarts.edu',
      logout: jest.fn()
    }
    expect(renderer.create(<AppUI {...props}/>).toJSON()).toMatchSnapshot()
  })
  it('renders a test page with a no logged in user but an email present the same as no logged in user', () => {
    const props: AppProps = {
      page: <div> Test Page </div>,
      isLoggedIn: false,
      userEmail: 'snape@hogwarts.edu',
      logout: jest.fn()
    }
    expect(renderer.create(<AppUI {...props}/>).toJSON()).toMatchSnapshot()
  })
})
