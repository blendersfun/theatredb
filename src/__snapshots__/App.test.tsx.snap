// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logic initializes the state when constructed 1`] = `<div />`;

exports[`ui renders a test page with a no logged in user but an email present the same as no logged in user 1`] = `
<div
  className="App"
>
  <div
    className="navMenu"
  >
    <div
      className="navLinks"
    >
      <a
        href="#page=home"
      >
        Home
      </a>
      <a
        href="#page=data-entry"
      >
        Data Entry
      </a>
    </div>
    <div
      className="loginStatus"
    >
      <a
        href="#page=login"
      >
        Login
      </a>
    </div>
  </div>
  <div>
     Test Page 
  </div>
</div>
`;

exports[`ui renders a test page with a user logged in 1`] = `
<div
  className="App"
>
  <div
    className="navMenu"
  >
    <div
      className="navLinks"
    >
      <a
        href="#page=home"
      >
        Home
      </a>
      <a
        href="#page=data-entry"
      >
        Data Entry
      </a>
    </div>
    <div
      className="loginStatus"
    >
      <div>
        <span
          className="whoami"
        >
          Logged in as 
          snape@hogwarts.edu
        </span>
        <a
          href="javascript:;"
          onClick={[MockFunction]}
        >
          Logout
        </a>
      </div>
    </div>
  </div>
  <div>
     Test Page 
  </div>
</div>
`;

exports[`ui renders a test page with a user not logged in 1`] = `
<div
  className="App"
>
  <div
    className="navMenu"
  >
    <div
      className="navLinks"
    >
      <a
        href="#page=home"
      >
        Home
      </a>
      <a
        href="#page=data-entry"
      >
        Data Entry
      </a>
    </div>
    <div
      className="loginStatus"
    >
      <a
        href="#page=login"
      >
        Login
      </a>
    </div>
  </div>
  <div>
     Test Page 
  </div>
</div>
`;
